Python/ceval.c

```diff
@@ -2005,6 +2008,48 @@ _PyEvalFramePushAndInit(PyThreadState *tstate, PyFunctionObject *func,
     return NULL;
 }
 
+/* Same as _PyEvalFramePushAndInit but takes an args tuple and kwargs dict.
+   Steals references to func, callargs and kwargs.
+*/
+static _PyInterpreterFrame *
+_PyEvalFramePushAndInit_Ex(PyThreadState *tstate, PyFunctionObject *func,
+    PyObject *locals, Py_ssize_t nargs, PyObject *callargs, PyObject *kwargs)
+{
+    bool has_dict = (kwargs != NULL && PyDict_GET_SIZE(kwargs) > 0);
+    PyObject *kwnames = NULL;
+    PyObject *const *newargs = has_dict
+        ? _PyStack_UnpackDict(tstate, _PyTuple_ITEMS(callargs),
+            nargs, kwargs, &kwnames)
+        : &PyTuple_GET_ITEM(callargs, 0);
+    if (newargs == NULL) {
+        Py_DECREF(func);
+        goto error;
+    }
+    if (!has_dict) {
+        /* We need to incref all our args since the new frame steals the references. */
+        for (Py_ssize_t i = 0; i < nargs; ++i) {
+            Py_INCREF(PyTuple_GET_ITEM(callargs, i));
+        }
+    }
+    _PyInterpreterFrame *new_frame = _PyEvalFramePushAndInit(
```

Comment:
Rather than transforming `callargs` and `kwargs` into a shape suitable for `_PyEvalFramePushAndInit`, it might make sense to push the frame, then unpack the tuple and dict into the the frame without the intermediate objects.
It complicates the errror handling a bit, and the frame will need to be cleaned up if there is an error.

Reply:
I cam work on this in another PR.